Sources/GRPC/GRPCTLSConfiguration.swift

```diff
@@ -0,0 +1,603 @@
+/*
+ * Copyright 2021, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import NIOSSL
+
+#if canImport(Network)
+import Network
+import NIOTransportServices
+import Security
+#endif
+
+/// TLS configuration.
+///
+/// This structure allow configuring TLS for a wide range of TLS implementations. Some
+/// options are removed from the user's control to ensure the configuration complies with
+/// the gRPC specification.
+public struct GRPCTLSConfiguration {
+  fileprivate enum Backend {
+    /// Configuration for NIOSSSL.
+    case nio(NIOConfiguration)
+    #if canImport(Network)
+    /// Configuration for Network.framework.
+    case network(NetworkConfiguration)
+    #endif
+  }
+
+  /// The TLS backend.
+  private var backend: Backend
+
+  private init(backend: Backend) {
+    self.backend = backend
+  }
+
+  /// Return the configuration for NIOSSL or `nil` if Network.framework is being used as the
+  /// TLS backend.
+  internal var nioConfiguration: NIOConfiguration? {
+    switch self.backend {
+    case let .nio(configuration):
+      return configuration
+    case .network:
+      return nil
+    }
+  }
+
+  internal var isNetworkFrameworkTLSBackend: Bool {
+    switch self.backend {
+    case .nio:
+      return false
+    #if canImport(Network)
+    case .network:
+      return true
+    #endif
+    }
+  }
+
+  /// The server hostname override as used by the TLS SNI extension.
+  ///
+  /// This value is ignored when the configuration is used for a server.
+  internal var hostnameOverride: String? {
+    get {
+      switch self.backend {
+      case let .nio(config):
+        return config.hostnameOverride
+
+      #if canImport(Network)
+      case let .network(config):
+        return config.hostnameOverride
+      #endif
+      }
+    }
+
+    set {
+      switch self.backend {
+      case var .nio(config):
+        config.hostnameOverride = newValue
+        self.backend = .nio(config)
+
+      #if canImport(Network)
+      case var .network(config):
+        if #available(macOS 10.14, iOS 12.0, watchOS 5.0, tvOS 12.0, *) {
+          if let hostnameOverride = newValue {
+            config.updateHostnameOverride(to: hostnameOverride)
+          } else {
+            // We can't unset the value so error instead.
```

Comment:
In principle we could allow unsetting by lazily setting the value (i.e. set it at usage time). For now this is probably fine, but it may be worth documenting the caveat.

Reply:
Good idea! I added FIXME and documented in the relevant places.